osquery/tables/system/hash.cpp

```diff
@@ -139,7 +158,108 @@ std::string hashFromFile(HashType hash_type, const std::string& path) {
   }
 }
 
-namespace tables {
+/*
+ * FileHashCache implements persistent in-memory caching for files' hashes.
+ * this cache has LRU eviction policy. the hash is recalculated
+ * every time the mtime or size of the file changes.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#if defined(WIN32)
+
+typedef struct _stat Stat;
+#define xstat(path, st) _stat((path), static_cast<struct _stat *>(st))
+
+#else
+
+#include <unistd.h>
+
+typedef struct stat Stat;
+#define xstat(path, st) stat((path), static_cast<struct stat *>(st))
+
+#endif
+
+struct FileHashCache {
```

Comment:
Considering you're moving this new `struct` into the osquery `namespace`, would this not be better implemented as a custom class? Seems like this is a great candidate for RIAA, and you might get some better flexibility with operators and it's usage down below.

Reply:
I started like that, but then I realised it wasn't worth the complexity increase. This type of cache is very specific to this application, so I'd rather leave it like this. If any use case for similar caching arises I'm glad and willing to refactor it. I can move it outside of the namespace if it would make it less confusing (although it is gonna increase the code verbosity).